### PR TITLE
WIP: Remove mac from xml network def

### LINKED ISSUE #396: unable to create a static ipv6 routed domain
#  Version Reports:

### Distro  version of host:

``` CentOS Linux release 7.5.1804 (Core) ```

### Terraform Version Report

``` Terraform v0.11.8 ```

### Libvirt version

```3.9.0```

### terraform-provider-libvirt plugin version (git-hash)

``` f104fad4```
___
# Description of Issue/Question

### Setup
```
provider "libvirt" {
    uri = "qemu:///system"
}

resource "libvirt_network" "public"{
  name = "public"
  mode = "route"
  bridge = "br-public"
  addresses = ["{{ ipv6_network }}"]
  dhcp {
    enabled = "false"
  }
  autostart = "true"
}

resource "libvirt_domain" "tf_test" {
  # init
  name = "tf_test"
  metadata = "tf_test"
  vcpu = 2
  memory = 512
  running = false
  autostart = false

  # network
  network_interface {
    network_id = "${libvirt_network.public.id}"
    addresses = ["{{ ipv6_address }}"]
    # wait_for_lease = true
  }
}
```

### Steps to Reproduce Issue
```
terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + libvirt_domain.tf_test
      id:                               <computed>
      arch:                             <computed>
      autostart:                        "false"
      emulator:                         <computed>
      machine:                          <computed>
      memory:                           "512"
      metadata:                         "tf_test"
      name:                             "tf_test"
      network_interface.#:              "1"
      network_interface.0.addresses.#:  "1"
      network_interface.0.addresses.0:  "{{ ipv6_address }}"
      network_interface.0.hostname:     <computed>
      network_interface.0.mac:          <computed>
      network_interface.0.network_id:   "${libvirt_network.public.id}"
      network_interface.0.network_name: <computed>
      qemu_agent:                       "false"
      running:                          "false"
      vcpu:                             "2"

  + libvirt_network.public
      id:                               <computed>
      addresses.#:                      "1"
      addresses.0:                      "{{ ipv6_network }}"
      autostart:                        "true"
      bridge:                           "br-public"
      dhcp.#:                           "1"
      dhcp.0.enabled:                   "false"
      mode:                             "route"
      name:                             "public"


Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes
  
  libvirt_network.public: Creating...
  addresses.#:    "" => "1"
  addresses.0:    "" => "{{ ipv6_network }}"
  autostart:      "" => "true"
  bridge:         "" => "br-public"
  dhcp.#:         "" => "1"
  dhcp.0.enabled: "" => "false"
  mode:           "" => "route"
  name:           "" => "public"
libvirt_network.public: Creation complete after 5s (ID: 3cc28126-de07-4cbc-b4ff-cd62c234cc38)
libvirt_domain.tf_test: Creating...
  arch:                             "" => "<computed>"
  autostart:                        "" => "false"
  emulator:                         "" => "<computed>"
  machine:                          "" => "<computed>"
  memory:                           "" => "512"
  metadata:                         "" => "tf_test"
  name:                             "" => "tf_test"
  network_interface.#:              "" => "1"
  network_interface.0.addresses.#:  "" => "1"
  network_interface.0.addresses.0:  "" => "{{ ipv6_address }}"
  network_interface.0.hostname:     "" => "<computed>"
  network_interface.0.mac:          "" => "<computed>"
  network_interface.0.network_id:   "" => "3cc28126-de07-4cbc-b4ff-cd62c234cc38"
  network_interface.0.network_name: "" => "<computed>"
  qemu_agent:                       "" => "false"
  running:                          "" => "false"
  vcpu:                             "" => "2"

Error: Error applying plan:

1 error(s) occurred:

* libvirt_domain.tf_test: 1 error(s) occurred:

* libvirt_domain.tf_test: virError(Code=27, Domain=19, Message='XML error: Invalid to specify MAC address '76:92:e8:5a:a4:9b' in network 'public' IPv6 static host definition')
```
If I try to uncomment `wait_for_lease = true` the domain is corrected created, terraform doesn't end because `dchp enabled is false`, so no lease will be assigned to the domain

### PR DESCRIPTION
**Pull request by MalloZup**
# What does this PR?

In this Pr when when we create a new network interface  we don't assign a random generate MAC address anymore on the xml-definition.
Instead we delegate it to libvirt which autogenerate it randomly.

- remove of code for mac adress generation.
- adapt to wait_for_lease 

## addition:
- this pr  additionaly fix `wait_for_lease` in doc and in test which was `wait_for_lease` = 1 but we need a boolean `wait_for_lease = true` 

# Previous behaviour:
In past we were assigning a random mac adress to the XML definition.
This doesn"t play well with IPV6 see (Fix #396)

## Libvirt upstream doc
___
from https://avdv.github.io/libvirt/formatnetwork.html
____
mac
    The address attribute defines a MAC (hardware) address formatted as 6 groups of 2-digit hexadecimal numbers, the groups separated by colons (eg, "52:54:00:1C:DA:2F"). This MAC address is assigned to the bridge device when it is created. **Generally it is best to not specify a MAC address when creating a network** - in this case, if a defined MAC address is needed for proper operation, **libvirt will automatically generate a random MAC address** and save it in the config. Allowing libvirt to generate the MAC address will assure that it is compatible with the idiosyncrasies of the platform where libvirt is running. 

# What is left todo:

- [x] check other removal needed

- [x] tests network setups

- [x] finish check wait_for_lease mechanism.

libvirt will autogenerate a proper one.

____
special thx to @lgaggini

### CODE REVIEW

**Comment by MalloZup** [comment]
for this PR special thx for this to @lgaggini which helped a lot :cupid:  thx.

So all the testacc are running fine.
I have tested the 2 main.tf from example issue, and other network configuration  and everything seems working.

I cannot see why we need to inject/create a mac

**Comment by MalloZup** [comment]
ok the tests are fine. feel free to have a look @inercia @dmacvicar . i will do more tests on this but looks good to me now

File: libvirt/network_def_test.go [hunk 1]
```diff
@@ -50,8 +49,8 @@ func TestNetworkDefUnmarshall(t *testing.T) {
 			<ip address="192.168.122.1" netmask="255.255.255.0">
 				<dhcp>
 					<range start="192.168.122.100" end="192.168.122.254" />
-					<host mac="00:16:3e:77:e2:ed" name="foo.example.com" ip="192.168.122.10" />
-					<host mac="00:16:3e:3e:a9:1a" name="bar.example.com" ip="192.168.122.11" />
+					<host name="foo.example.com" ip="192.168.122.10" />
```
Comment (dmacvicar) [inline libvirt/network_def_test.go]: I think I understand the removal above, as it fits the description of why we remove the bridge address, but I don't understand why we remove the matching in the DHCP host list.

**Comment by MalloZup** [comment]
@dmacvicar ok thx for feedback, i will rearrange it.
at moment i need to fix the `wait_for_lease`  mechanism.

File: libvirt/network_def_test.go (hunk 1 above)
Comment (MalloZup) [reply libvirt/network_def_test.go]: the mac here is only testing the unmarshalling mechanism as unit-test. 
We can leave it, yop.

File: libvirt/domain.go [hunk 2]
```diff
@@ -78,31 +78,22 @@ func domainWaitForLeases(domain *libvirt.Domain, waitForLeases []*libvirtxml.Dom
 	return err
 }
 
-func domainIfaceHasAddress(domain libvirt.Domain, iface libvirtxml.DomainInterface,
-	domainDef libvirtxml.Domain, virConn *libvirt.Connect, rd *schema.ResourceData) (found bool, ignore bool, err error) {
+func domainIfaceHasAddress(domain libvirt.Domain, domainDef libvirtxml.Domain, virConn *libvirt.Connect, rd *schema.ResourceData) (found bool, ignore bool, err error) {
 
-	mac := strings.ToUpper(iface.MAC.Address)
```
Comment (MalloZup) [inline libvirt/domain.go]: why this removal:

with the new behaviour we are assigning an empty MAC to network interfaces.

The `iface` here is the temporary interface.
https://github.com/dmacvicar/terraform-provider-libvirt/pull/422/files#diff-2f03be069459ebc91fcf0607d4d686d8R1316

With `domainGetIfacesInfo(domain, domainDef, virConn, rd)`  we get the `real` interfaces so we have the MAC adresses there.

**Comment by MalloZup** [comment]
Currently with patch:
this works:

```hcl
provider "libvirt" {
  uri = "qemu:///system"
}

resource "libvirt_domain" "domain" {
  name = "domain-${count.index}"
  network_interface {
    network_name = "default"
    wait_for_lease = true
  }

  count = 1
}
      
```

We get the `IP` after waiting for lease correctly.

this **won't work** ( previous behaviour was working)
```hcl
provider "libvirt" {
    uri = "qemu:///system"
}

resource "libvirt_network" "public4"{
  name = "public4"
  mode = "route"
  bridge = "br-public4"
  addresses = ["10.0.1.0/24"]
  dhcp {
    enabled = "false"
  }
  autostart = "true"
}

resource "libvirt_domain" "tf_test4" {
  # init
  name = "tf_test"
  metadata = "tf_test"
  vcpu = 2
  memory = 512
  running = false
  autostart = false

  network_interface {
    network_id = "${libvirt_network.public4.id}"
    addresses = ["10.0.1.3"]
    wait_for_lease = true
  }
}

```

we cannot get the DomainInterface infos and we loop until the `5min` go over.
```golang
2018-09-20T12:59:42.885+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] qemu-agent is not used
2018-09-20T12:59:42.885+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] getting domain addresses from networks
2018-09-20T12:59:42.886+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] Interfaces: ([]libvirt.DomainInterface) <nil>
2018-09-20T12:59:42.886+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] no interfaces could be obtained with qemu-agent: falling back to the libvirt API
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] Interfaces info obtained with libvirt API:
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: ([]libvirt.DomainInterface) {
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: }
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: 
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] ifaces with addresses: []
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] doesn't have IP address(es) yet...
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [DEBUG] IP address not found for interface  &{XMLName:{Space: Local:} Managed: TrustGuestRXFilters: MAC:0xc420215090 Source:0xc420204540 Boot:<nil> VLan:<nil> VirtualPort:<nil> IP:[] Route:[] Script:<nil> Target:<nil> Guest:<nil> Model:0xc420215000 Driver:<nil> Backend:<nil> FilterRef:<nil> Tune:<nil> Link:<nil> MTU:<nil> Bandwidth:<nil> Coalesce:<nil> ROM:<nil> Alias:<nil> Address:<nil>}  will try in a while
2018-09-20T12:59:42.887+0200 [DEBUG] plugin.terraform-provider-libvirt: 2018/09/20 12:59:42 [TRACE] Waiting 10s before next try
```

I think i am missing something :thinking:  but i cannot figure find out.. any hints welcome :+1:

**Comment by dmacvicar** [comment]
You have _DHCP_ disabled, which means you want to get an address from the DHCP behind the bridge.

* I don't think adding `addresses = ["10.0.1.3"]` has any effect, as you don't control that DHCP server host table, and you can't choose what address it should give to you.
* _wait_for_lease_ loops over the function that retrieves domain ip addresses. This is done primarily by _virDomainInterfaceAddresses_ using the leases file, which only works if you manage the network inside _libvirt_ and _libvirt_ has access to the leases file.
* This may work if we enable _ARP_ as source, as I suggested in PR #427 
* This may work with the current qemu agent code, if the agent is running in the image (and also if we refactor to use _virDomainInterfaceAddresses_ using _AGENT_ as _src_)

**Comment by MalloZup** [comment]
i will return to work on this soon.

@dmacvicar  yop this is also what i was thinking, and i think this is true.
I have some concern of some part of the code like the partialNetIface retrieval.

I will need to test and check some small part of code let see the findings.

Thx :+1: :heart:

**Comment by MalloZup** [comment]
@dmacvicar  feel free to review it. I double checked it more and i think we did it wrong even with the past.

I did a test with `master` and `this` branch. ( from the above tf. files)
for example  with the master branch a `virsh dumpxml domaintest` give
```
  <interface type='network'>
      <mac address='ee:6f:4e:b9:65:47'/>
      <source network='public4'/>
      <model type='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
    </interface>
```
and with the new branch gives : 

```
      <interface type='network'>
        <mac address='52:54:00:1b:71:d5'/>
        <source network='public4' bridge='br-public4'/>
        <target dev='vnet0'/>
        <model type='virtio'/>
        <alias name='net0'/>
        <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
      </interface>
```
So we have a bridge network tag in the `nomac` tag is correct. I dunno why in the master we don't have it.

Anyway i think the PR add the correct behaviour, even if the `wait_for_lease` variable is somehow not easy to understand a first glance ( and also is not 100% right for QEMU-agent..) 

So just double check it : tia :love_hotel:

**Comment by Bischoff** [comment]
@dmacvicar :
> I don't think adding addresses = ["10.0.1.3"] has any effect  (without DHCP)

It has. It sets the address on the bridge.

The whole point of recent PRs was precisely to decouple these addresses and DHCP.

**Comment by MalloZup** [comment]
@Bischoff feel free to check this out, i will turn back on this once i got the qemu-agent PR merged ( should be done also soon).

**Comment by MalloZup** [comment]
Closing I will reopen properly in future. Parkinh prs aren't fun imho🦄

**Pull request closed without merging**